Improve our warning standing w.r.t. unused lambda captures

blozano-tt opened the pull request:
### Ticket
#23444 

### Problem description
We have this compiler warning disabled.
`unused-lambda-captures`

This has allowed for an immense amount of dead code surrounding lambdas in our code base.

The problem is at its worst in `ttnn`.

### What's changed
- Fix violations in `tt_stl`, `tt_metal`, and `tests`
- Continue to ignore the warning in `ttnn` and `tt-train`

### Checklist
- [x] [All post commit](https://github.com/tenstorrent/tt-metal/actions/runs/18480952438) CI passes
- [x] New/Existing tests provide coverage for changes

blozano-tt committed:
Improve our warning standing w.r.t. unused lambda captures

blozano-tt committed:
Update tt-train/CMakeLists.txt

blozano-tt commented:
/codeowners ping

athompsonTT commented on third_party/CMakeLists.txt:
those are probably my fault, I went overboard capturing `this` to fix a different capture warning.  I'll look into fixing the patch file to not capture `this` when its not used.

athompsonTT commented on tt_metal/programming_examples/distributed/3_distributed_eltwise_add/distributed_eltwise_add.cpp:
```diff
@@ -122,7 +122,7 @@ int main() {
 
     // Create and initialize source data
-    constexpr float val_to_add = 0.5f;
+    constexpr auto val_to_add = 0.5f;
```
was this necessary?

athompsonTT committed:
TT-Train: clean up lambda captures (#30614)

### Ticket
N/A

### Problem description
TT-Train had some lambdas that were over-capturing.

### What's changed
Remove unused captures.
Remove warning supression.

### Checklist
- [ ] [All post
commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml)
CI passes

blozano-tt commented:
/codeowners ping

blozano-tt committed:
Merge branch 'main' into blozano-unused-lambda-capture

blozano-tt commented:
/codeowners ping

blozano-tt committed:
Merge branch 'main' into blozano-unused-lambda-capture
